### PR TITLE
DeveloperAccount and DeveloperUser CRDs

### PR DESCRIPTION
**Pull request by eguzki**
CRD (and controller) for managing 3scale developer accounts.
CRD (and controller) for managing 3scale developer user accounts.

Requires https://github.com/3scale/3scale-porta-go-client/pull/40

- [x] CR samples
- [x] Usage doc / reference doc
- [x] CRD diagram update
- [x] User Suspend/Unsuspend
- [x] User Password from secret
- [x]  validation of user email

### CODE REVIEW

**Comment by eguzki** [comment]
@miguelsorianod ready for initial review. Only doc is missing.

File: pkg/helper/boolean_utils.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package helper
+
+func Any(arr []bool) bool {
+	result := false
+	for _, v := range arr {
+		result = result || v
```
Comment (miguelsorianod) [inline pkg/helper/boolean_utils.go:6]: Is this intended to check whether at least one of the bools in the array is `true`? if so, we can just check it and return without needing to iterate all the array.
Same would happen with `All` but just checking the opposite.

File: apis/capabilities/v1beta1/developeraccount_types.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+const (
+	DeveloperAccountKind = "DeveloperAccount"
```
Comment (miguelsorianod) [inline apis/capabilities/v1beta1/developeraccount_types.go:35]: As a side comment, conceptually this should really belong to one package level above as it's independent of the API version but we can leave it here

File: apis/capabilities/v1beta1/developeruser_types.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	// Username
+	Username string `json:"username"`
+
+	// Email
+	Email string `json:"email"`
+
+	// Password
+	PasswordCredentialsRef corev1.SecretReference `json:"passwordCredentialsRef"`
+
+	// DeveloperAccountRef is the reference to the parent developer account
```
Comment (miguelsorianod) [inline apis/capabilities/v1beta1/developeruser_types.go:76]: This means we only support specifying developer accounts that have been created through the capabilities functionality right?

File: controllers/capabilities/developeraccount_status_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+		ID:                  s.resource.Status.ID,
+		AccountState:        s.resource.Status.AccountState,
+		CreditCardStored:    s.resource.Status.CreditCardStored,
+		ProviderAccountHost: s.resource.Status.ProviderAccountHost,
+		Conditions:          s.resource.Status.Conditions.Copy(),
+		ObservedGeneration:  s.resource.Status.ObservedGeneration,
+	}
+
+	// If there is an error and s.remoteDeveloperAccount is nil
+	// ID is not changed, as it is initialized with previous values d
```
Comment (miguelsorianod) [inline controllers/capabilities/developeraccount_status_reconciler.go:90]: Is the comment incomplete?

I' trying to understand this use case

File: controllers/capabilities/developeraccount_threescale_reconciler.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+	desiredMonthlyBillingEnabled := true
+	if s.resource.Spec.MonthlyBillingEnabled != nil {
+		desiredMonthlyBillingEnabled = *s.resource.Spec.MonthlyBillingEnabled
+	}
+	if devAccount.Element.MonthlyBillingEnabled != nil && *devAccount.Element.MonthlyBillingEnabled != desiredMonthlyBillingEnabled {
+		update = true
+		deltaAccount.Element.MonthlyBillingEnabled = &desiredMonthlyBillingEnabled
+	}
+
+	// MonthlyChargingEnabled defaults to False
```
Comment (miguelsorianod) [inline controllers/capabilities/developeraccount_threescale_reconciler.go:186]: Is the comment outdated or is the value incorrect?

**Review by miguelsorianod — COMMENTED** [review]
It looks good :+1: 

I left some comments

File: pkg/helper/boolean_utils.go (hunk 1 above)
Comment (eguzki) [reply pkg/helper/boolean_utils.go:6]: right, this is an alternative implementation. But both approaches work, right?

File: apis/capabilities/v1beta1/developeruser_types.go (hunk 3 above)
Comment (eguzki) [reply apis/capabilities/v1beta1/developeruser_types.go:76]: Right. 

Would be more flexible to make reference to Account ID? I think we should be generic and make references to other CR names or unique fields in the CR (like system_name in other CRD's). But I do not have strong opinion.

File: controllers/capabilities/developeraccount_status_reconciler.go (hunk 4 above)
Comment (eguzki) [reply controllers/capabilities/developeraccount_status_reconciler.go:90]: This comment does not make sense here, the message was added few lines above. 

The use case is that `status.ID` has been set correctly. In later reconciliation loop, there is some unexpected error and `s.remoteDeveloperAccount` (the 3scale developer account object read from the API) can be nil. Therefore, the `status.ID` is not changed to nil, instead, the existing value remains untouched.

File: controllers/capabilities/developeraccount_threescale_reconciler.go (hunk 5 above)
Comment (eguzki) [reply controllers/capabilities/developeraccount_threescale_reconciler.go:186]: fixed

**Comment by eguzki** [comment]
:+1: I am going to write the doc

File: pkg/helper/boolean_utils.go (hunk 1 above)
Comment (miguelsorianod) [reply pkg/helper/boolean_utils.go:6]: both should work yes. I was pointing there was no need to iterate all array

File: apis/capabilities/v1beta1/developeruser_types.go (hunk 3 above)
Comment (miguelsorianod) [reply apis/capabilities/v1beta1/developeruser_types.go:76]: I don't have a strong opinion about it either.

File: config/crd/patches/webhook_in_developerusers.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# The following patch enables conversion webhook for CRD
+# CRD conversion requires k8s 1.13 or later.
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (miguelsorianod) [inline config/crd/patches/webhook_in_developerusers.yaml:3]: The APIVersion i for `webhook_*` files and `cainjection_*` files should be `v1`

**Review by miguelsorianod — CHANGES_REQUESTED** [review]
Doc looks good to me :+1: .

There's one pending change regarding APIversion of some files

File: config/crd/patches/webhook_in_developerusers.yaml (hunk 6 above)
Comment (eguzki) [reply config/crd/patches/webhook_in_developerusers.yaml]: fixed

**Review by miguelsorianod — APPROVED** [review]

**Pull request merged**